java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java

```diff
@@ -405,7 +405,12 @@ static ValueRange getValueRange(ColumnStatistics index,
       TimestampColumnStatistics stats = (TimestampColumnStatistics) index;
       Timestamp min = useUTCTimestamp ? stats.getMinimumUTC() : stats.getMinimum();
       Timestamp max = useUTCTimestamp ? stats.getMaximumUTC() : stats.getMaximum();
-      return new ValueRange<>(predicate, min, max, stats.hasNull());
+      // To support predicates with ns precision we widen min and max ms range by 1 to ensure ns values are in range.
```

Comment:
This is the wrong pattern, unfortunately. You'll need to follow the example of what was done for the string shortening. (We just store the first 1024 characters for min/max on strings.) The problem comes that if someone does a test that looks for equality, this change will return wrong results.

Reply:
Thanks for the comments @omalley ! Latest patch updates TimestampStatistics to store the lower 6 digits (that were initially ignored) in order to achieve nanosecond precision.

Also liked the idea proposed by @csringhofer and added 0 and 999_999 as DEFAULT_NANOS MIN and MAX respectively when these values are missing.

Please let me know what you think!